airbyte-cdk/python/airbyte_cdk/sources/streams/checkpoint/checkpoint_reader.py

```diff
@@ -94,67 +94,147 @@ def __init__(self, cursor: Cursor, stream_slices: Iterable[Optional[Mapping[str,
+                state_for_slice = self._cursor.select_state(next_slice)
                 if state_for_slice == {"__ab_full_refresh_sync_complete": True}:
-                    return None
+                    next_candidate_slice = None
+                    has_more = True
+                    while has_more:
+                        next_candidate_slice = self.get_next_slice()
+                        state_for_slice = self._cursor.select_state(next_candidate_slice)
+                        has_more = state_for_slice == {"__ab_full_refresh_sync_complete": True}
+                    return StreamSlice(cursor_slice=state_for_slice or {}, partition=next_candidate_slice.partition)
                 else:
-                    return self._current_slice
-            if self._read_state_from_cursor:
+                    return StreamSlice(cursor_slice=state_for_slice or {}, partition=next_slice.partition)
+            else:
                 state_for_slice = self._cursor.select_state(self._current_slice)
                 if state_for_slice == {"__ab_full_refresh_sync_complete": True}:
-                    self._current_slice = self._get_next_slice()
+                    # Skip every slice that already has the terminal complete value indicating that a previous attempt
+                    # successfully synced the slice
+                    next_candidate_slice = None
+                    has_more = True
+                    while has_more:
+                        next_candidate_slice = self.get_next_slice()
+                        state_for_slice = self._cursor.select_state(next_candidate_slice)
+                        has_more = state_for_slice == {"__ab_full_refresh_sync_complete": True}
+                    return StreamSlice(cursor_slice=state_for_slice or {}, partition=next_candidate_slice.partition)
                 else:
-                    self._current_slice = StreamSlice(cursor_slice=state_for_slice or {}, partition=self._current_slice.partition)
-            else:
-                # Unlike RFR cursors that iterate dynamically based on how stream state is updated, most cursors operate on a
-                # fixed set of slices determined before reading records. They should just iterate to the next slice
-                self._current_slice = self._get_next_slice()
-            return self._current_slice
-        except StopIteration:
-            self._finished_sync = True
-            return None
+                    return StreamSlice(cursor_slice=state_for_slice or {}, partition=self._current_slice.partition)
+        else:
+            # Unlike RFR cursors that iterate dynamically according to how stream state is updated, most cursors operate
+            # on a fixed set of slices determined before reading records. They just iterate to the next slice
+            return self.get_next_slice()
 
-    def _get_next_slice(self) -> StreamSlice:
+    def get_next_slice(self) -> StreamSlice:
```

Comment:
@girarda ah yeah i think this is a product of it just all being private originally so probably paid less attention to the naming but  it makes sense to name this better. I'll just call it `read_and_convert_slice()`. The key thing is that its supposed to get the next slice and then convert it into a `StreamSlice` if it is not already of that type.

I guess it comes down to preference. It was originally private because it was internal to `_find_next_slice()`. However, because we now have the `LegacyCursorBasedCheckpointReader` class which is a subclass of `CursorBasedCheckpointReader`, this method needed to be overridable since the legacy one needs to read and convert the mapping to a `StreamSlice`. And since we're now overriding I made this method public, but the shared private code in `_find_next_slice()` is meant to be shared and not overwritten.